mlflow/tracing/trace_manager.py

```diff
@@ -0,0 +1,177 @@
+import logging
+import threading
+from dataclasses import dataclass, field
+from functools import lru_cache
+from typing import Dict, Optional
+
+from opentelemetry import trace as trace_api
+
+from mlflow.tracing.types.model import Status, StatusCode, Trace, TraceData, TraceInfo
+from mlflow.tracing.types.wrapper import MLflowSpanWrapper, NoOpMLflowSpanWrapper
+
+_logger = logging.getLogger(__name__)
+
+
+# Internal representation to keep the state of a trace.
+# Dict[str, Span] is used instead of TraceData to allow access by span_id.
+@dataclass
+class _Trace:
+    trace_info: TraceInfo
+    span_dict: Dict[str, MLflowSpanWrapper] = field(default_factory=dict)
+
+    def to_mlflow_trace(self) -> Trace:
+        span_list = [span.to_mlflow_span() for span in self.span_dict.values()]
+        return Trace(self.trace_info, TraceData(spans=span_list))
+
+
+class InMemoryTraceManager:
+    """
+    Manage spans and traces created by the tracing system in memory.
+    """
+
+    _instance_lock = threading.Lock()
+    _instance = None
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance is None:
+            with cls._instance_lock:
+                if cls._instance is None:
+                    cls._instance = InMemoryTraceManager()
+        return cls._instance
+
+    def __init__(self):
+        self._traces: Dict[str, _Trace] = {}
+        self._lock = threading.Lock()  # Lock for _traces
+
+    def start_detached_span(
+        self,
+        name: str,
+        trace_id: Optional[str] = None,
+        parent_span_id: Optional[str] = None,
+        span_type: Optional[str] = None,
+    ) -> MLflowSpanWrapper:
```

Comment:
Can we add span_id here? Is it possible to create a span with a known span_id in opentelemetry?

Reply:
UnfortunatelyIt's not possible, trace_id and span_id are uuid generated by them.